manager/allocator/allocator_new.go

```diff
@@ -0,0 +1,993 @@
+					log.G(ctx).WithError(err).Error("task cannot be allocated")
+				}
+			}
+		}
+
+		// now handle all of the tasks belonging to no service
+		for _, task := range tasks[""] {
+			ctx := log.WithField(ctx, "task.id", task.ID)
+			allocTaskDone := metrics.StartTimer(allocatorActions.WithValues("task", "allocate"))
+			err := a.network.AllocateTask(task)
+			allocTaskDone()
+			switch {
+			case err == nil:
+				// first, if no error occurred, then we can add this to the
+				// list of tasks we will commit
+				allocatedTasks = append(allocatedTasks, task)
+			case errors.IsErrAlreadyAllocated(err):
+				// if the task is already allocated, there is nothing to do. log
+				// that and do not add this network to the batch
+				log.G(ctx).Debug("task already fully allocated")
+			case errors.IsErrRetryable(err):
+				// if the error is retryable, then we should log that error
+				// and readd the task to the pending tasks map
+				log.G(ctx).WithError(err).Error("task could not be allocated, but will be retried")
+				a.pendingTasks[task.ID] = struct{}{}
+			default:
+				// default covers any other error case. specifically, error that
+				// are not retryable. for these, we should fail and not re-add them
+				// to the pending map. they will never succeed. additionally,
+				// remove these tasks from the map
+				log.G(ctx).WithError(err).Error("task cannot be allocated")
+			}
+		}
+
+		// finally, if we have any services or tasks to commit, open a
+		// batch and commit them
+		if len(allocatedServices)+len(allocatedTasks) > 0 {
+			taskWriteTxDone := metrics.StartTimer(storeLockHeld.WithValues("allocate", "task", "write"))
+			if err := a.store.Batch(func(batch *store.Batch) error {
+				for _, service := range allocatedServices {
+					if err := batch.Update(func(tx store.Tx) error {
+						currentService := store.GetService(tx, service.ID)
+						if currentService == nil {
+							// NOTE(dperny): see explanation in node
+							log.G(ctx).WithField("service.id", service.ID).Debug("service was delete after allocation but before commit")
+							a.network.DeallocateService(service)
+							a.deletedObjects[service.ID] = struct{}{}
+							return nil
+						}
+						// the service may have changed in the meantime, so
+						// only update the fields we just altered in it.
+						// in this case, only the endpoint
+						currentService.Endpoint = service.Endpoint
+						// then commit the service
+						return store.UpdateService(tx, currentService)
+					}); err != nil {
+						// TODO(dperny): i'm unsure how to handle batch
+						// failures, other than maybe crashing the allocator.
+						// Batch covers up what nodes actually succeeded and
+						// what failed.
+					}
+				}
+
+				for _, task := range allocatedTasks {
+					if err := batch.Update(func(tx store.Tx) error {
+						currentTask := store.GetTask(tx, task.ID)
+						if currentTask == nil ||
+							currentTask.Status.State != api.TaskStateNew ||
```

Comment:
Should this be `>=api.TaskStateNew`?  Otherwise if it's in `api.TaskStatePending` or `api.TaskStateRunning` (which shouldn't happen if the above logic is correct for adding tasks to the pending map, but it might be good to be conservative), we'd deallocate.

Reply:
I don't think it helps to be conservative. If we end up in a situation where a task has been updated to task state `PENDING` but is also being allocated, we're pretty much in a bad state already.